server: adds truncated exponential backoff + rand for retry

cfromknecht opened the pull request:
This commit adds a backoff policy to the peer termination watcher to avoid getting stuck in tight connection loops with failing peers. The maximum backoff is currently set to 128s, and each backoff is randomized so that two instances using the same algorithm have some hope of desynchronizing. Note, a peer's backoff will only be reset if we manually disconnect via `DisconnectPeer` or restart the server.

Roasbeef commented on lntest/harness.go:
```diff
@@ -281,6 +281,7 @@ func (n *NetworkHarness) ConnectNodes(ctx context.Context, a, b *HarnessNode) er
 			Pubkey: bobInfo.IdentityPubkey,
 			Host:   b.cfg.P2PAddr(),
 		},
+		Perm: true,
```
Not sure if this should be set to perm, as there're conditions where we want to connect peers in the tests, but then after a restart they shouldn't attempt to automatically reconnect.

aakselrod replied:
Agreed here.

cfromknecht replied:
I mainly added this to be sure this doesn't get us stuck in any weird states, the itests seem to handle it just fine. Will revert.

Roasbeef commented on server.go:
```diff
@@ -42,6 +44,14 @@ var (
+
+	// defaultBackoff is the starting point for exponential backoff for
+	// reconnecting to persistent peers.
+	defaultBackoff = time.Millisecond
```
Seems too low for an initial starting value?

aakselrod replied:
Agreed, 1s might be a good starting point.

cfromknecht replied:
Originally had it at 1 second but was failing itests even without perm. Swapping back now seems to be fine tho lol

Roasbeef commented on server.go:
```diff
@@ -42,6 +44,14 @@ var (
+
+	// maximumBackoff is the largest backoff we will permit when
+	// reattempting connections to persistent peers.
+	maximumBackoff = 128 * time.Second
```
Why not just use `time.Minute` here instead?

cfromknecht replied:
I like base 2 and I cannot lie

aakselrod commented on server.go:
```diff
@@ -1232,24 +1253,39 @@ func (s *server) peerTerminationWatcher(p *peer) {
-
-		// If so, then we'll attempt to re-establish a persistent
-		// connection to the peer.
+		// Otherwise, we'll launch a new connection requests in order to
```
Nit: "request" should be singular.

cfromknecht commented:
Thank you @aakselrod and @Roasbeef for the reviews, I've updated the default backoff to 1s, the maximum backoff to 1m, corrected the grammatical error, and removed the use of Perm from the lntest package! ⚡️

Roasbeef commented:
Can you rebase this? Thanks!
